[NNC] Add python bindings for Compute2

huiguoo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59411 [NNC] Fix the uninitialized pointer in loopnest.fuse_loops
* **#59350 [NNC] Add python bindings for Compute2**
* #59348 [NNC] Fix BufHandle arguments in loopnest python API

Differential Revision: [D28854806](https://our.internmc.facebook.com/intern/diff/D28854806)

huiguoo committed:
[NNC] Add python bindings for Compute2

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28854806).

navahgar commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -223,6 +223,19 @@ void initTensorExprBindings(PyObject* module) {
       py::return_value_policy::reference);
+
+  te.def(
+      "Compute2",
```
Why not add this as another case under the def above for `Compute`?

huiguoo replied:
As the case of `dim_args.size()>4`? It seems unnecessary (`dim_args.size()>4`) to use `Compute2`-like API (function argument being a vector). Also, we may not want the user to check the number of `dim_args` before they use `Compute`.

navahgar replied:
I didn't mean for the user to check the number of `dim_args`. The user will just pass a `func` to `Compute`. That `func` either takes in 1, 2, 3, or 4 parameters (as is currently the case) or takes in a vector (which is the case you are trying to add) if the number of args is > 4.

Are you suggesting that the user might want to use this API even when the number of args is <= 4?

I am just trying to see if having a new function name `Compute2` does not create more confusion.

ZolotukhinM replied:
I think it could be confusing that for 4 arguments I'd need to pass a function that takes 4 arguments and for 5 arguments I'd need to pass a function that takes one argument (a list with 5 elements). From that perspective having a separate function, like `Compute2` makes sense to me since I don't see a way to distinguish these two cases inside `Compute`. That said, I think `Compute2` could use a better name.

navahgar replied:
Actually any function that special cases for smaller number of arguments will do that, say, for example, have overloaded functions for 1, 2, 3 and 4 arguments, and another overload that takes a list (for args > 4). That is pretty normal IMO. Having a different method is more confusing for me. 

But if you both prefer this approach, I don't have a problem.

ZolotukhinM replied:
Mm, I'd like to have the overloaded function here too, but I'm not sure it's possible because of the fact that we're using `py::function` here and that's what in theory should be used to resolve the overload. In other words, we cannot distinguish when a user passes a function that expects a list and a function that expects 4 parameters. And since we can not distinguish that, we cannot resolve the overload, and hence we need to have explicitly different functions for these two cases. Am I missing something? I just don't see how this would work otherwise: if I provide a function that accepts a list and then happen to pass a list with 2 elements, then the current implementation will try to expand the list into two parameters and call the function with two separate parameters, which will not work.

navahgar replied:
Oh, yeah, okay. I see the problem with this approach now. Thanks for clarifying.

navahgar reviewed (approved):
LGTM

huiguoo committed:
Update on "[NNC] Add python bindings for Compute2"


Differential Revision: [D28854806](https://our.internmc.facebook.com/intern/diff/D28854806)

[ghstack-poisoned]

huiguoo commented:
@huiguoo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28854806).
